ably/realtime_presence.go

```diff
@@ -127,82 +194,147 @@ func (pres *RealtimePresence) SyncComplete() bool {
 	if pres.syncState == syncInProgress {
 		return
 	} else if pres.syncState != syncInitial {
 		// Sync has started, make all callers to Get(true) wait. If it's channel's
 		// initial sync, the callers are already waiting.
 		pres.syncMtx.Lock()
 	}
-	pres.serial = serial
 	pres.syncState = syncInProgress
-	pres.stale = make(map[string]struct{}, len(pres.members))
-	for memberKey := range pres.members {
-		pres.stale[memberKey] = struct{}{}
+	pres.beforeSyncMembers = make(map[string]*PresenceMessage, len(pres.members)) // RTP19
+	for memberKey, member := range pres.members {
+		pres.beforeSyncMembers[memberKey] = member
+	}
+}
+
+// RTP19, RTP19a
+func (pres *RealtimePresence) leaveMembers(members map[string]*PresenceMessage) {
+	for memberKey := range members {
+		delete(pres.members, memberKey)
+	}
+	for _, msg := range members {
+		msg.Action = PresenceActionLeave
+		msg.ID = ""
+		msg.Timestamp = time.Now().UnixMilli()
+		pres.messageEmitter.Emit(msg.Action, (*subscriptionPresenceMessage)(msg))
 	}
 }
 
 func (pres *RealtimePresence) syncEnd() {
 	if pres.syncState != syncInProgress {
 		return
 	}
-	for memberKey := range pres.stale {
-		delete(pres.members, memberKey)
-	}
-	for memberKey, presence := range pres.members {
+	pres.leaveMembers(pres.beforeSyncMembers) // RTP19
+
+	for memberKey, presence := range pres.members { // RTP2f
 		if presence.Action == PresenceActionAbsent {
 			delete(pres.members, memberKey)
 		}
 	}
-	pres.stale = nil
+	pres.beforeSyncMembers = nil
 	pres.syncState = syncComplete
 	// Sync has completed, unblock all callers to Get(true) waiting
 	// for the sync.
 	pres.syncMtx.Unlock()
 }
 
-func (pres *RealtimePresence) processIncomingMessage(msg *protocolMessage, syncSerial string) {
-	for _, presmsg := range msg.Presence {
-		if presmsg.Timestamp == 0 {
-			presmsg.Timestamp = msg.Timestamp
+// RTP2a, RTP2b, RTP2c
+func (pres *RealtimePresence) addPresenceMember(memberMap map[string]*PresenceMessage, memberKey string, presenceMember *PresenceMessage) bool {
+	if existingMember, ok := memberMap[memberKey]; ok { // RTP2a
+		isMemberNew, err := presenceMember.IsNewerThan(existingMember) // RTP2b
+		if err != nil {
+			pres.log().Error(err)
+			errorInfo := newError(0, err)
+			pres.channel.errorEmitter.Emit(subscriptionName("error"), (*errorMessage)(errorInfo))
 		}
+		if isMemberNew {
+			memberMap[memberKey] = presenceMember
+			return true
+		}
+		return false
+	}
+	memberMap[memberKey] = presenceMember
+	return true
+}
+
+// RTP2a, RTP2b, RTP2c
+func (pres *RealtimePresence) removePresenceMember(memberMap map[string]*PresenceMessage, memberKey string, presenceMember *PresenceMessage) bool {
+	if existingMember, ok := memberMap[memberKey]; ok { // RTP2a
+		isMemberNew, err := presenceMember.IsNewerThan(existingMember) // RTP2b
+		if err != nil {
+			pres.log().Error(err)
+			errorInfo := newError(0, err)
+			pres.channel.errorEmitter.Emit(subscriptionName("error"), (*errorMessage)(errorInfo))
```

Comment:
as discussed above -- I'd get rid of this

Reply:
Not sure why we even added this in the first place. I am sure there was a concrete reason

Reply:
You can see in `ably-dotnet`, we have a `realtimechannel` method called `OnError`.
https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L559-L567
This gets called whenever there is any error registered on a channel 
https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L628

You can see there is an error delegate registration here 
https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L57C44-L57C44

Reply:
`OnError` method usages
1. Presence -> https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/Presence.cs#L610
2. Sync message -> https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/ChannelMessageProcessor.cs#L148
3. Channel state errors -> https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L628

Reply:
Let me know if we should keep it. If this is not part of the spec, then there's no point in keeping the public method.

Reply:
Seems we need to do something like this 
https://github.com/ably/ably-java/blob/72992c943498ee2ebefb8acec303718ed7f35ae4/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L1336C1-L1339C6

Reply:
Okay, I have removed this particular piece of code. 
Let me know if we planning to add `OnError` method to the spec. I think this will be a very specific way to listen to channel-specific errors as we do the same in `ably-dotnet`.

Reply:
> You can see in ably-dotnet, we have a realtimechannel method called OnError.

I am not a dot.net developer. But my understanding is that if it's not part of the realtime channel interface then it's not part of the public api. It's an internal method, used by the library itself. It's annotated 'public' because it can be called by other parts of the library, but that doesn't make it part of the actual public API.

> Let me know if we planning to add OnError method to the spec.

We're not.

> [Seems we need to do something like this](https://github.com/ably/ably-java/blob/72992c943498ee2ebefb8acec303718ed7f35ae4/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L1336C1-L1339C6)

[We already have that](https://github.com/ably/ably-go/blob/fix/presencemap/ably/realtime_channel.go#L995)

Reply:
Got it 👍